plot residual

HanHsuanWu opened the pull request:
### Description of the change
A few sentences and/or a bulleted list to describe and motivate the change:

As suggested by @thomasaarholt on gitter, I added a plot_residual argument to model1d.plot. 
This argument will plot a residual signal (Signal - Model) to the signal plot.
Currently, it has a bug when the size of signal and model are not the same. I am not sure what's the best way to do this so any help is appreciated.

### Progress of the PR
- [x] Change implemented (can be split into several points),
- [x] update docstring (if appropriate),
- [ ] update user guide (if appropriate),
- [x] add an changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/upcoming_changes/README.rst)),
- [x] Check formatting changelog entry in the `readthedocs` doc build of this PR (link in github checks)
- [ ] add tests,
- [ ] ready for review.

### Minimal example of the bug fix or the new feature
```python
import hyperspy.api as hs
import numpy as np
s = hs.signals.Signal1D(np.tile(np.arctan(np.arange(-500, 500)), (10, 1)),dtype='float64')
s.add_gaussian_noise(std=0.1)
m = s.create_model()
line  = hs.model.components1D.Polynomial(order=2)
m.append(line)
m.fit()
m.plot(plot_residual=True)
```

HanHsuanWu committed:
plot residual

thomasaarholt commented on hyperspy/models/model1d.py:
```diff
@@ -710,15 +710,36 @@ def _model2plot(self, axes_manager, out_of_range2nans=True):
+    
+    def _residual2plot(self, axes_manager, out_of_range2nans=True):
+        old_axes_manager = None
+        if axes_manager is not self.axes_manager:
```
Could you document what happens here?

HanHsuanWu replied:
I added some comment for what _residual_for_plot does.

HanHsuanWu replied:
I have added documentation for _residual_for_plot.

thomasaarholt commented on hyperspy/models/model1d.py:
```diff
@@ -710,15 +710,36 @@ def _model2plot(self, axes_manager, out_of_range2nans=True):
-
-    def plot(self, plot_components=False, **kwargs):
+    
+    def _residual2plot(self, axes_manager, out_of_range2nans=True):
```
Hyperspy does use the `foo2bar` syntax here and there, but I think its generally considered better to use `foo_to_bar`. So here I'd suggest `_residual_to_plot`, or perhaps better `_residual_for_plot` and `out_of_range_to_nan`.

thomasaarholt commented on hyperspy/models/model1d.py:
```diff
@@ -710,15 +710,36 @@ def _model2plot(self, axes_manager, out_of_range2nans=True):
+        if old_axes_manager is not None:
+            self.axes_manager = old_axes_manager
+            self.fetch_stored_values()
+        if out_of_range2nans is True:
```
Since there currently is no way to change the value of `out_of_range2nans` from a calling function, I suggest you just remove the `if` and always use the nan-behaviour.

If you do keep the `if`, then you can remove `is True`, since `if out_of_range2nans` will evaluate to `if True` or `if False`.

thomasaarholt commented on hyperspy/models/model1d.py:
```diff
@@ -742,6 +763,21 @@ def plot(self, plot_components=False, **kwargs):
         self._connect_parameters2update_plot(self)
+        
+        #Optional to plot the residual of (Signal - Model)
+        if plot_residual is True:
```
`if plot_residual is True:` -> `if plot_residual:`

thomasaarholt reviewed (commented):
Thanks for the PR! I've added some general linting comments below as well as a few suggestions.

HanHsuanWu commented:
Hey @thomasaarholt thanks for reviewing.
I have addressed your comments and ended up deleting the out_of_range2nans part.
I added a new argument for model1d.__call__() to ignore the channel_switches, so that it always plots the residual for the whole spectrum. We can add another argument to turn this on or off if needed. What do you think?

HanHsuanWu committed:
Update model1d.py

ericpre commented on hyperspy/models/model1d.py:
```diff
@@ -710,15 +718,36 @@ def _model2plot(self, axes_manager, out_of_range2nans=True):
+        s = (self.signal.__call__() - self.__call__(ignore_channel_switches = True))
+        if old_axes_manager is not None:
+            self.axes_manager = old_axes_manager
+            self.fetch_stored_values()
```
This can be simplify to:

```suggestion
    def _residual_for_plot(self):
        """From an model1D object, the original signal is subtracted
        by the model signal then returns the residual
        """

        return self.signal.__call__() - self.__call__(ignore_channel_switches=True)

```
Github doesn't allow me to make a suggestion over a deleted line and `return s` below obviously also need to be removed in the above suggestion.

HanHsuanWu replied:
I added a keyword argument to _residual_for_plot to prevent the error message attached below.
def _residual_for_plot(self,**kwargs):
![image](https://github.com/hyperspy/hyperspy/assets/112028355/8ff58b08-823b-4aba-897e-adfe5755532f)

ericpre commented on hyperspy/models/model1d.py:
```diff
@@ -413,7 +418,10 @@ def __call__(self, non_convolved=False, onlyactive=False,
+            if not ignore_channel_switches:
+                axis = self.axis.axis[self.channel_switches]
+            else:
+                axis = self.axis.axis
```
A more compact version:
```suggestion
            slice_ = slice(None) if ignore_channel_switches else self.channel_switches
            axis = self.axis.axis[slice_]
```

ericpre reviewed (commented):
Thank you @HanHsuanWu, this works well and is a long overdue!

I have suggested some simplification.

This would need a changelog entry.

HanHsuanWu committed:
Update model1d.py

Co-authored-by: Eric Prestat <eric.prestat@gmail.com>

HanHsuanWu commented:
> Thank you @HanHsuanWu, this works well and is a long overdue!
> 
> I have suggested some simplification.
> 
> This would need a changelog entry.

What is a changelog entry and where do I add this?

ericpre commented:
See https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/upcoming_changes/README.rst as per checklist of the first comment of the PR.
It is a note to describe the changes of this PR and will be added automatically to the list of changes (also called changelog) for each release.

HanHsuanWu committed:
added changelog

HanHsuanWu commented:
> See https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/upcoming_changes/README.rst as per checklist of the first comment of the PR. It is a note to describe the changes of this PR and will be added automatically to the list of changes (also called changelog) for each release.

Thanks Eric. I have added the changelog. Got a bit confused before.

HanHsuanWu commented:
@ericpre any further comments for this pull request?

ericpre commented:
Sorry, I don't have time at the minute, I will come back to it very soon.

ericpre committed:
Add test and example for plot_residual

ericpre reviewed (approved):
Thanks @HanHsuanWu, this looks good.
In 18a019a8c7db0ce36768d997b49af14a8c548588 I added a test and an example. I have also rebased on the `RELEASE_next_major` branch as the next release will be the 2.0.
